--- upstream_new/src/fixed-dtoa.h	2011-11-05 20:57:46.147349597 +0100
+++ upstream_new/src/fixed-dtoa.h	2011-10-31 17:41:34.102313165 +0100
@@ -51,7 +51,7 @@
 // This method only works for some parameters. If it can't handle the input it
 // returns false. The output is null-terminated when the function succeeds.
 bool FastFixedDtoa(double v, int fractional_count,
-                   Vector<char> buffer, int* length, int* decimal_point);
+                   Vector<char> buffer, int* length, int* decimal_point, DoubleToStringConverter::RoundingBias rounding_bias);

 #ifndef NO_DOUBLE_CONVERSION_NAMESPACE
 }  // namespace double_conversion
