--- upstream_unpatched/src/strtod.cc	2012-06-08 10:19:24.575018959 +0200
+++ upstream/src/strtod.cc	2012-06-08 10:33:19.015013238 +0200
@@ -508,7 +508,9 @@
   double double_previous = Double(double_guess).PreviousDouble();
 
   float f1 = static_cast<float>(double_previous);
+#ifdef DEBUG_ENABLE_OPASSERT
   float f2 = float_guess;
+#endif // DEBUG_ENABLE_OPASSERT
   float f3 = static_cast<float>(double_next);
   float f4;
   if (is_correct) {
